src/features/bivariate_manager/components/ConnectedBivariateMatrix/ConnectedBivariateMatrix.tsx

```diff
@@ -37,151 +39,161 @@ const mapHeaderCell = (
-        ? yGroups.findIndex(
-            (group) =>
-              group.selectedQuotient[0] === matrixSelection?.yNumerator &&
-              group.selectedQuotient[1] === matrixSelection?.yDenominator,
-          )
-        : -1;
-
-      return { x: xIndex, y: yIndex };
-    }, [xGroups, yGroups, matrixSelection]);
-
-    const selectedCell = useRef<{ x: number; y: number }>(
-      calculateSelectedCell(),
-    );
-
-    useEffect(() => {
-      selectedCell.current = calculateSelectedCell();
-    }, [matrixSelection, xGroups, yGroups]);
-
-    const headings = useMemo(() => {
+const ConnectedBivariateMatrix = forwardRef<HTMLDivElement | null, any>(({}, ref) => {
+  const [matrixSelection, { setMatrixSelection }] = useAtom(bivariateMatrixSelectionAtom);
+
+  const [matrix] = useAtom(bivariateCorrelationMatrixAtom);
+
+  const [{ xGroups, yGroups }, { setNumerators }] = useAtom(bivariateNumeratorsAtom);
+  const setTooltip = useAction(currentTooltipAtom.setCurrentTooltip);
+  const turnOffById = useAction(currentTooltipAtom.turnOffById);
+  const resetTooltip = () => turnOffById(TOOLTIP_ID);
+
+  const [{ data: statisticsData }] = useAtom(bivariateStatisticsResourceAtom);
+
+  const stats = statisticsData?.polygonStatistic.bivariateStatistic;
+
+  const calculateSelectedCell = useCallback((): { x: number; y: number } => {
+    const xIndex = xGroups
+      ? xGroups.findIndex(
+          (group) =>
+            group.selectedQuotient[0] === matrixSelection?.xNumerator &&
+            group.selectedQuotient[1] === matrixSelection?.xDenominator,
+        )
+      : -1;
+    const yIndex = yGroups
+      ? yGroups.findIndex(
+          (group) =>
+            group.selectedQuotient[0] === matrixSelection?.yNumerator &&
+            group.selectedQuotient[1] === matrixSelection?.yDenominator,
+        )
+      : -1;
+
+    return { x: xIndex, y: yIndex };
+  }, [xGroups, yGroups, matrixSelection]);
+
+  const selectedCell = useRef<{ x: number; y: number }>(calculateSelectedCell());
+
+  useEffect(() => {
+    selectedCell.current = calculateSelectedCell();
+  }, [matrixSelection, xGroups, yGroups]);
+
+  const headings = useMemo(() => {
+    if (
+      !stats ||
+      !stats.indicators ||
+      !xGroups ||
+      !xGroups.length ||
+      !yGroups ||
+      !yGroups.length
+    ) {
+      return null;
+    }
+
+    const mapWithIndicators = (axis: 'x' | 'y') => (group: AxisGroup) =>
+      mapHeaderCell(group, stats?.indicators, stats?.correlationRates, axis);
+
+    return {
+      x: xGroups.map(mapWithIndicators('x')),
+      y: yGroups.map(mapWithIndicators('y')),
+    };
+  }, [stats, xGroups, yGroups]);
+
+  const debouncedShowTooltip = useCallback(
+    debounce((position) => {
+      resetTooltip();
+      setTooltip({
+        popup: <ProgressTooltip close={resetTooltip} />,
+        position,
+        initiatorId: TOOLTIP_ID,
+        hoverBehavior: true,
+      });
+    }, 400),
+    [resetTooltip, setTooltip],
+  );
```

Comment:
Can you explain why do you use debounce?

Reply:
same reason of using in inputs, here we can select cells/heading fast, we don't need this tooltip on every select, just in the end.